### PR TITLE
Decompile viewport management

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openloco/window.h [hunk 1]
```diff
@@ -250,6 +275,7 @@ namespace openloco::ui
         }
 
         static viewport_pos map_from_3d(loc16 loc, int32_t rotation);
+        void centre_2d_coordinates(int16_t x, int16_t y, int16_t z, int16_t* outX, int16_t* outY);
```
Comment (ZehMatt) [inline src/openloco/window.h:289]: Perhaps it makes more sense to return a struct or pair of int16_t.

File: src/openloco/viewportmgr.cpp [hunk 2]
```diff
@@ -1,32 +1,128 @@
...
+    static viewport* init_viewport(gfx::point_t origin, gfx::ui_size_t size, ZoomLevel zoom)
+    {
+        auto vp = _viewports.emplace_back(std::make_unique<viewport>()).get();
+
+        vp->x = origin.x;
+        vp->y = origin.y;
+        vp->width = size.width;
+        vp->height = size.height;
+
+        vp->view_width = size.width << (uint8_t)zoom;
```
Comment (ZehMatt) [inline src/openloco/viewportmgr.cpp]: Consider using static_cast to ensure the cast is valid.

File: src/openloco/viewportmgr.cpp [hunk 3]
```diff
@@ -1,32 +1,128 @@
...
+        vp->view_height = size.height << (uint8_t)zoom;
+        vp->zoom = (uint8_t)zoom;
```
Comment (ZehMatt) [inline src/openloco/viewportmgr.cpp]: Perhaps we could directly use the enum here?

File: src/openloco/viewportmgr.cpp [hunk 4]
```diff
@@ -70,16 +168,243 @@ namespace openloco::ui::viewportmgr
...
+            rect.left <<= viewport->zoom;
+            rect.top <<= viewport->zoom;
+            rect.right <<= viewport->zoom;
+            rect.bottom <<= viewport->zoom;
+
+            if (!viewport->intersects(rect))
+                continue;
+
+            // Get intersection
+            auto left = std::max(rect.left, viewport->view_x);
```
Comment (ZehMatt) [inline src/openloco/viewportmgr.cpp]: Could also use ViewportRec here.

File: src/openloco/viewportmgr.cpp [hunk 5]
```diff
@@ -70,16 +168,243 @@ namespace openloco::ui::viewportmgr
...
+            top = top >> viewport->zoom;
+            bottom = bottom >> viewport->zoom;
+
+            // offset calculated area by viewport offset
+            left += viewport->x;
+            right += viewport->x;
+            top += viewport->y;
+            bottom += viewport->y;
+
+            gfx::set_dirty_blocks(left, top, right, bottom);
```
Comment (ZehMatt) [inline src/openloco/viewportmgr.cpp:267]: Could also use the ViewportRect as argument, clarifies that the coordinates are screen space.

File: src/openloco/window.cpp [hunk 6]
```diff
@@ -348,7 +348,7 @@ namespace openloco::ui
         int32_t base_height = map::tile_element_height(*map_x, *map_y);
         int16_t dest_x, dest_y;
         viewport* v = this->viewports[0];
-        centre_2d_coordinates(*map_x, *map_y, base_height, &dest_x, &dest_y, v);
+        v->centre_2d_coordinates(*map_x, *map_y, base_height, &dest_x, &dest_y);
 
         // Rebase mouse position onto centre of window, and compensate for zoom level.
         int16_t rebased_x = ((this->width >> 1) - mouse_x) * (1 << v->zoom),
```
Comment (ZehMatt) [inline src/openloco/window.cpp:354]: Might be also worth to add a utility function that scales values based on zoom, this kind of code will probably pop up else where.

File: src/openloco/viewportmgr.cpp [hunk 7]
```diff
@@ -1,32 +1,128 @@
...
+            vp->var_12 |= (1 << 5);
         }
 
-        _viewportPointers[0] = nullptr;
+        return vp;
+    }
+
+    static void focusViewportOn(window* w, int index, thing_id_t dx)
+    {
+        viewport* viewport = w->viewports[index];
```
Comment (ZehMatt) [inline src/openloco/viewportmgr.cpp:70]: Could need an assert for debug builds to test index.

File: src/openloco/viewportmgr.cpp (hunk 5 above)
Comment (marijnvdwerf) [reply src/openloco/viewportmgr.cpp:267]: I'd like to have ViewportRect separate from ScreenRect. Viewports are in "rendered world" space, while the screen is like... a layer on top of that. If that makes sense.

**Comment by marijnvdwerf** [comment]
I agree with the other comments, but would like to get this merged without too many code restructurings. We really need types + structs/classes for different coordinates.

**Pull request merged**